### ISSUE TITLE
Can't fetch liquid from any tank

### ISSUE THREAD

**Issue by martinhoite** [McJty/RFToolsControl#33]
So I've been struggling with this program for a few days now - I finally decided to go full debug mode.

No matter what I do, I can't get the processor to fetch liquids (and as such I can't test fetch liquid).

I've gone as far as to _literally_ copy Direwolf20's example in his spotlight, and changed the sides, facing, block placing, tried with a node, nothing works.

I can read amounts and max in tanks just fine, and also what type of liquid, but fetch does nothing.

Tested with: Tinker's Construct Tanks (multiblock, and the normal ones), EnderIO and a Deep Resonance tank.

Current setup:
<img src="http://imgur.com/czotvfL.png" />
<img src="http://imgur.com/3QmyYIq.png" />

<s>Going to test if I can move items now.</s>
I can move items from one chest to another just fine, so seems to be only liquids.

Additional info:
Running a server with Direwolf20 1.10 pack (version 1.0.0)
RFtools control version: 1.5.2

**Comment by Unadvoidable** [comment]
Can you copy your code into a txt file? (Programmer> Ctrl +a,  ctrl+c, tab out of game and paste into txt file)

ps. Did you allocate the liquid slot within the processor and put that variable into the InSlot?

**Comment by McJty** [comment]
yes, copy/paste the code here

**Comment by martinhoite** [comment]
Ah, so that's how you get it onto pastebin etc. :)

[LiquidTest.txt](https://github.com/McJty/RFToolsControl/files/632068/LiquidTest.txt)

That's the simple program I used for testing, the current configuration is for a EnderIO tank - I've changed the sides to all possible variants, so that shouldn't be the issue.

> Did you allocate the liquid slot within the processor and put that variable into the InSlot?

Absolutely, I double checked everything and I can't find anything wrong
![2016-12-05_20 23 54](https://cloud.githubusercontent.com/assets/5943291/20898988/ccbdd114-bb28-11e6-945c-ec958bf75912.png)

**Comment by McJty** [comment]
I've had trouble with the enderio tanks myself I must say. I'm not sure they support capabilities correctly as all other tanks seem to work fine. Maybe try to report to them? I know extra utilities drums and Deep Resonance tanks work fine

**Comment by martinhoite** [comment]
Here's the weird thing, Direwolf20 used that exact setup in his mod spotlight, and I've seen several reports off the Tinker's Construct new multiblock tanks working just fine, and I can't get ANY of them to work (I'll go test drums now though)

Update:
Just tried every single drum from Extra Utilities as well, none of them work either.
Once again, I can read the values just fine, but I can't fetch liquids, not water, not lava, not oil, none of them, regardless of the used tank.

**Comment by Unadvoidable** [comment]
I just tested your small program:
In this case its the Multitank causing problems, it refuses to accept fluids N/W/S/E - U/D are valid though.

Maybe the Multitank index for North South East West is shifted - It is always recognized correctly in the slot thing, thats why it doesn't throws an error but makes error afterwards ? That could alsow explain my issiue with moving liquid crystal somehow ....

Man the code is futuretech rocketsience but could it maybe be an followup mistake of 
_GuiProcessor.java - line 527_ ?
` EnumFacing side = EnumFacing.values()[i / TANKS];`

The thing with Ender I/O tanks is to make sure to select a certain side an make shure it is configured to all but Pull or Disabled.

**Workaround**
for this simple situation -> Place Multitank above or below

![image](https://cloud.githubusercontent.com/assets/4508203/20900854/24ae854c-bb30-11e6-8d06-711cc3562041.png)

**Comment by martinhoite** [comment]
Even though I was certain I had tried moving the tank around, your workaround does indeed work - thanks!

But there's still something not working quite as intended.

**Comment by McJty** [comment]
You are only supposed to access multitanks by the internal slots. Not directly with fluid opcodes

**Comment by martinhoite** [comment]
I think we're talking past each other here McJty.

If you look at the simple program linked above ([LiquidTest](https://github.com/McJty/RFToolsControl/files/632068/LiquidTest.txt)), I'm not trying to access the multitank in any other way than what is intended  unless I misunderstood something completely.

The goal of the program, is to grab fluid from a normal tank, and transfer it to the multitank.

After that, the goal would be to transfer the liquid from the multitank to another tank, using opcodes.

To recap what the issue _appears_ to be:

I _can't_ fetch liquid from a normal tank (like a drum etc.) into the multitank, and thereby cannot push it to another tank.
Unadvoidable posted the following workaround (which works):

> for this simple situation -> Place Multitank above or below

But isn't the multitank supposed to be able to be placed on **ANY** side of the processor?
If yes, then _that_ is the issue I'm reporting, if no - ignore this, and keep it closed.

Regards Blizz

**Comment by McJty** [comment]
Ok, but your program is confusing me. Can you explain to me exactly how you are putting liquids in and out of the tank and multitanks? What opcodes with what parameters?

**Comment by martinhoite** [comment]
Right, I removed everything that isn't necessary to explain the issue :)

First program does **NOT** work
[MultiTankOnSouthSide.txt](https://github.com/McJty/RFToolsControl/files/634182/MultiTankOnSouthSide.txt)
Setup:
![multitanksouth](https://cloud.githubusercontent.com/assets/5943291/20930491/ce44fb98-bbcd-11e6-9dbe-884cf141ec06.png)

Second program **DOES** work
[MultiTankOnUpSide.txt](https://github.com/McJty/RFToolsControl/files/634184/MultiTankOnUpSide.txt)
Setup:
![multitankabove](https://cloud.githubusercontent.com/assets/5943291/20930513/e5a943e8-bbcd-11e6-8fd6-25fc7e471df4.png)


So to followup on the workaround from before:
If the multitank is **not** above, _or_ below the processor, the program will not work

It's the exact same program in both cases, only difference is the side where the tanks are (so also changed in the fetch and push opcodes)

To clarify: The multitank should works just fine on _any_ adjacent side of the processor, correct?

Hope that helps clear it up!

**Comment by McJty** [comment]
In the first program (the not working one) is the multitank on the west side?

**Comment by McJty** [comment]
Also when changing the program/setup did you redo the allocation for the fluid slots in the processor?

**Comment by martinhoite** [comment]
No, it's on the south side:
![2016-12-06_16 11 18](https://cloud.githubusercontent.com/assets/5943291/20930707/cbba0d2c-bbce-11e6-92d3-5c9efd3a3c43.png)

West side is where the button is

And to answer your other question, no, I didn't - the allocation has been the same the entire time, and I started with the Multi Tank on the south side

**Comment by McJty** [comment]
Let me try some things

**Comment by martinhoite** [comment]
Opcodes in the programmers

Fetch:
![2016-12-06_16 17 28](https://cloud.githubusercontent.com/assets/5943291/20930925/9298078c-bbcf-11e6-970f-f423f1200af6.png)

Push:
![2016-12-06_16 14 44](https://cloud.githubusercontent.com/assets/5943291/20930822/34cbb554-bbcf-11e6-89da-52340c31c4f9.png)

Edit: Oops, wrong order (fixed now)
Edit2: wrong bloody screenshot fixed again...

**Comment by McJty** [comment]
The problem is the slot allocation. After moving the multitank you need to reallocate. However, there seems to be a bug in rftools control where it is not updating the fluid slots. Fluid slots for a tank on top are actually internally different from fluid slots from a tank on another side. To fix this you have to extract/reinsert the cpu core to force it to update. Then you'll also see the allocation is no longer correct. Updating that should fix this. I'll fix the bug in rftools control though. It doesn't seem to notice when a multitank is moved

**Comment by martinhoite** [comment]
Took the CPU core out, fluid slots were still allocated, put it back in, still allocated.

Took the core out again, manually deselected the fluid allocations and reselected them, put the core back in.

Program still doesn't work, let me try moving the tank below without making any changes to the processor

**Comment by martinhoite** [comment]
Moving (or taking out and putting back in) the CPU core seems to have no effect on the processor, at all.

And moving the multitank below or above the processor is 100% guaranteed to work so far, even with different programs (moving liquid to different sides of the processor etc.), and a multitude of different allocations to test.

Having the Multi Tank on the North/East/South/West side renders it useless.

**Comment by McJty** [comment]
Ok there is certainly something wrong there. I will check this out further

**Comment by McJty** [comment]
I know what's wrong...

**Comment by McJty** [comment]
Should be fixed next release (1.6.2, in a moment)
